sprs/src/sparse/linalg/bicgstab.rs

```diff
@@ -0,0 +1,391 @@
+                            return Ok(Box::new(solver));
+                        }
+                    }
+                }
+
+                // If we ran past our iteration limit, error, but still return results
+                Err(Box::new(solver))
+            }
+
+            /// Reset the reference direction `rhat` to be equal to `r`
+            /// to prevent a singularity in `1 / rho`.
+            pub fn soft_restart(&mut self) {
+                self.soft_restart_count += 1;
+                self.rhat = self.r.to_owned();
+                self.rho = self.err * self.err; // Shortcut to (&self.r).squared_l2_norm();
+                self.p = self.r.to_owned();
+            }
+
+            /// Recalculate the error vector from scratch using `a` and `b`.
+            pub fn hard_restart(&mut self) {
+                self.hard_restart_count += 1;
+                // Recalculate true error
+                self.r = &self.b - &(&self.a.view() * &self.x.view()).view();
+                self.err = (&self.r).l2_norm();
+                // Recalculate reference directions
+                self.soft_restart();
+                self.soft_restart_count -= 1; // Don't increment soft restart count for hard restarts
+            }
+
+            pub fn step(&mut self) -> $T {
+                self.i += 1;
+
+                // Gradient descent step
+                let v = &self.a.view() * &self.p.view();
+                let alpha = self.rho / ((&self.rhat).dot(&v));
+                let h = &self.x + &self.p.map(|x| x * alpha); // latest estimate of `x`
+
+                // Conjugate direction step
+                let s = &self.r - &v.map(|x| x * alpha); // s = A*h
+                let t = &self.a.view() * &s.view();
+                let omega = t.dot(&s) / &t.squared_l2_norm();
+                self.x = &h.view() + &s.map(|x| omega * x);
+
+                // Check error
+                self.r = &s - &t.map(|x| x * omega);
+                self.err = (&self.r).l2_norm();
+
+                // Prep for next pass
+                let rho_prev = self.rho;
+                self.rho = (&self.rhat).dot(&self.r);
+
+                // Soft-restart if `rhat` is becoming perpendicular to `r`.
+                if self.rho.abs() / (self.err * self.err)
+                    < self.soft_restart_threshold
+                {
+                    self.soft_restart();
+                } else {
+                    let beta = (self.rho / rho_prev) * (alpha / omega);
+                    self.p = &self.r
+                        + (&self.p - &v.map(|x| x * omega)).map(|x| x * beta);
+                }
+
+                self.err
+            }
+
+            /// Set the minimum value of `rho` to trigger a soft restart
+            pub fn with_restart_threshold(mut self, thresh: $T) -> Self {
+                self.soft_restart_threshold = thresh;
+                self
+            }
+
+            /// Iteration number
+            pub fn i(&self) -> usize {
```

Comment:
Should this be given a more descriptive name such as `iteration_number`?

Reply:
But I have not read the papers and I don't know if this is standard lingo

Reply:
It's not standard lingo, just me being lazy - updated it to `iteration_count` for consistency with the other `*_count` values stored by the solver